### PR TITLE
UCP/EP/WIREUP: Implement client/server connection establishment.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/ucp/wireup/select.c [hunk 1]
```diff
@@ -1091,3 +1089,38 @@ ucs_status_t ucp_wireup_select_aux_transport(ucp_ep_h ep,
...
+        md       = context->tl_mds[md_index].md;
+        ucs_assert(context->tl_mds[md_index].attr.cap.flags & UCT_MD_FLAG_SOCKADDR);
+
+        if (uct_md_is_sockaddr_accessible(md, &params->sockaddr, UCT_SOCKADDR_ACC_REMOTE)) {
+            /* TODO use score to prefer best tl rather than using first one */
+            *rsc_index_p = tl_id;
+            return UCS_OK;
+        }
+
+        ucs_debug("ms %s cannot reach %s", context->tl_mds[md_index].rsc.md_name,
```
Comment (miked-mellanox) [inline src/ucp/wireup/select.c]: /ms/md

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5736/ for details (Mellanox internal link).

File: src/ucp/core/ucp_listener.c [hunk 2]
```diff
@@ -54,7 +75,12 @@ static ssize_t ucp_listener_conn_request_callback(void *arg, const void *conn_pr
                                           &prog_id);
     }
 
-    return 0;
+    return 0; /* replying to client by wireup request instead of private data */
```
Comment (alinask) [inline src/ucp/core/ucp_listener.c]: following the removal of the reply cb, will need to return UCS_OK here.

File: src/ucp/wireup/wireup_ep.h [hunk 3]
```diff
@@ -30,15 +30,20 @@ enum {
...
+    uct_ep_h                  sockaddr_ep;   /**< Used for client-server wireup */
     ucp_rsc_index_t           aux_rsc_index; /**< Index of auxiliary transport */
     volatile uint32_t         pending_count; /**< Number of pending wireup operations */
     volatile uint32_t         flags;         /**< Connection state flags */
     uct_worker_cb_id_t        progress_id;   /**< ID of progress function */
 };
 
+typedef struct ucp_wireup_client_data {
+    ucp_err_handling_mode_t err_mode;
+    uint64_t                ep_uuid;
```
Comment (alinask) [inline src/ucp/wireup/wireup_ep.h]: could you please add here that /* packed worker address follows */ ?

File: src/ucp/core/ucp_listener.c (hunk 2 above)
Comment (yosefe) [reply src/ucp/core/ucp_listener.c]: yes, and also remove the reply argument and not pass callback on client side..

File: src/ucp/wireup/wireup_ep.h (hunk 3 above)
Comment (yosefe) [reply src/ucp/wireup/wireup_ep.h]: ok

File: src/ucp/wireup/wireup.c [hunk 4]
```diff
@@ -188,23 +188,45 @@ static void ucp_wireup_process_request(ucp_worker_h worker, const ucp_wireup_msg
...
+            }
+            ep->flags |= UCP_EP_FLAG_DEST_UUID_PEER;
+        } else {
+            ucs_assert(ep->flags & UCP_EP_FLAG_DEST_UUID_PEER);
+        }
+    } else {
+        /* Reply for a client-server connection (client side) */
+        ep = ucp_worker_ep_find(worker, msg->ep_uuid);
+        if (ep == NULL) {
+            ucs_trace("got coonection request with invalid ep_uuid 0x%"PRIx64,
```
Comment (alinask) [inline src/ucp/wireup/wireup.c]: typo - connection

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5741/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5746/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5755/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5758/ for details (Mellanox internal link).

**Pull request merged**